pkg/cli/setup.go

```diff
@@ -0,0 +1,297 @@
+	configManager *config.Manager
+}
+
+// NewSetup creates a new setup.
+func NewSetup(opts SetupOpts) *Setup {
+	return &Setup{
+		errout:        opts.Errout,
+		prompt:        opts.Prompt,
+		logger:        opts.Logger,
+		loginFlow:     opts.LoginFlow,
+		configManager: opts.ConfigManager,
+	}
+}
+
+// Configure triggers the setup flow.
+func (s *Setup) Configure(flags *SetupFlags, clusterURL string) error {
+	if err := flags.Resolve(); err != nil {
+		return err
+	}
+
+	// Create a Cluster and an HTTP client with the few information already available.
+	cluster := NewCluster(nil)
+	cluster.SetURL(clusterURL)
+	cluster.SetTLS(TLS{Insecure: flags.insecure})
+
+	httpOpts := []httpclient.Option{
+		httpclient.Timeout(5 * time.Second),
+		httpclient.Logger(s.logger),
+		httpclient.NoFollow(),
+	}
+	httpClient := httpclient.New(cluster.URL(), httpOpts...)
+
+	// If no TLS verify policy is explicitly provided, download the cluster's CA bundle.
+	if !flags.insecure && len(flags.caBundle) == 0 {
+		needsDCOSCABundle, err := s.needsDCOSCABundle(httpClient)
+		if err != nil {
+			return err
+		}
+		if needsDCOSCABundle {
+			flags.caBundle, err = s.downloadDCOSCABundle(cluster.URL(), httpOpts, !flags.noCheck)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Create a cert pool from the CA bundle PEM. The user is prompted for manual
+	// verification of the certificate authority, unless `--no-check` is passed.
+	var certPool *x509.CertPool
+	if len(flags.caBundle) > 0 {
+		var err error
+		certPool, err = s.decodePEMCerts(flags.caBundle, !flags.noCheck)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Create an HTTP client for login, it uses the appropriate TLS configuration.
+	httpOpts = append(httpOpts, httpclient.TLS(&tls.Config{
+		InsecureSkipVerify: flags.insecure,
+		RootCAs:            certPool,
+	}))
+	httpClient = httpclient.New(cluster.URL(), httpOpts...)
+
+	// Login to get the ACS token.
+	acsToken, err := s.loginFlow.Start(flags.loginFlags, httpClient)
+	if err != nil {
+		return err
+	}
+	cluster.SetACSToken(acsToken)
+	httpClient = httpclient.New(cluster.URL(), append(httpOpts, httpclient.ACSToken(acsToken))...)
+
+	// Read cluster ID from cluster metadata.
+	metadata, err := NewDCOSClient(httpClient).Metadata()
+	if err != nil {
+		return err
+	}
+
+	// Read cluster name from Mesos state summary.
+	if stateSummary, err := mesos.NewClient(httpClient).StateSummary(); err == nil {
+		cluster.SetName(stateSummary.Cluster)
+	} else {
+		// Fallback to cluster ID as cluster name.
+		cluster.SetName(metadata.ClusterID)
+	}
+
+	// Create the config for the given cluster and attach the cluster.
+	err = s.configManager.Save(cluster.Config(), metadata.ClusterID, flags.caBundle)
+	if err != nil {
+		return err
+	}
+	return s.configManager.Attach(cluster.Config())
+}
+
+// needsDCOSCABundle checks whether or not the cluster certificate is already trusted
+// by the sytem. This is done by making an HTTPS request to the cluster. This check
```

Comment:
There is actually no guarantee that `needsDCOSCABundle` makes a TLS request. If the `httpClient` isn't configured to try secure URL it will correctly assume that it is not necessary to download DC/OS CA Bundle.

Reply:
Indeed, I'll adapt the comment. Also for optimization purpose it might make sense to check the URL scheme and return early with false if it's `http`, WDYT?

Reply:
Will it work though if you're trying to setup a cluster in `disabled` security mode?

Reply:
What might be problematic there?